### PR TITLE
Per-rule version constraints min-version/max-version

### PR DESCRIPTION
**Pull request by mjambon**
Key points:
* Each rule now support Semgrep version constraints `min-version` and `max-version`. If the Semgrep version doesn't fit within these constraints, the rule is skipped. It is a requirement that skipping such a rule be not fatal.
* We're introducing a new kind of rule error which is a new kind of error within the existing type (I tried the alternative of having a dedicated field for version incompatibilities but it got too complicated).
* Invalid rules used to cause semgrep to abort after parsing the rules and to discard all errors except the first one (note to us all: never ever discard errors silently). I changed this behavior in semgrep-core and in osemgrep (`--experimental`) so that we complete the scan even if some rules are invalid. I'm not sure about the consequences. The comments I removed mentioned that we were matching pysemgrep's behavior which aborts as soon as it finds an invalid rule (still doesn't tell us why it would abort early, though) and I haven't scrutinized pysemgrep long enough to see if it still performs some rule validation and early abort.
* A new `skipped_rules` field is now returned by semgrep (JSON only). It has the same type as the one that was already exposed by semgrep-core. It is not exposed by pysemgrep (yet?), only osemgrep for now. Both disabled in osemgrep for now so as to match pysemgrep's output.
* The `errors` field is now populated with errors coming from invalid rules in both pysemgrep and osemgrep.
* A new error level `Info` was introduced, associated with exit code 0.

Uses https://github.com/returntocorp/semgrep-interfaces/pull/113

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by aryx** [comment]
fail CI

File: semgrep.opam [hunk 1]
```diff
@@ -6,7 +6,7 @@ synopsis:
 description: """
 Semgrep is like grep but for searching patterns at the AST level.
 
-For more information see https://semsgrep.dev
+For more information see https://semgrep.com
```
Comment (aryx) [inline semgrep.opam]: this actually redirect to semgrep.dev, so maybe simpler to keep semgrep.dev?

File: cli/src/semgrep/core_output.py [hunk 2]
```diff
@@ -33,8 +33,8 @@
 def _core_location_to_error_span(location: core.Location) -> out.ErrorSpan:
     return out.ErrorSpan(
         file=location.path,
-        start=out.PositionBis(line=location.start.line, col=location.start.col),
-        end=out.PositionBis(line=location.end.line, col=location.end.col),
+        start=location.start,
```
Comment (aryx) [inline cli/src/semgrep/core_output.py:37]: would have been better to have those changes in a separate PR, but ok.

File: src/core/Rule.ml [hunk 3]
```diff
@@ -587,18 +592,37 @@ and invalid_rule_error_kind =
...
-type error =
+type error_kind =
   | InvalidRule of invalid_rule_error
   | InvalidYaml of string * Tok.t
   | DuplicateYamlKey of string * Tok.t
   | UnparsableYamlException of string
 
-(* can't use Error because it's used for severity *)
-exception Err of error
+type error = {
```
Comment (aryx) [inline src/core/Rule.ml:609]: This could also be a separate PR

File: src/osemgrep/cli_scan/Cli_json_output.ml [hunk 4]
```diff
@@ -215,15 +224,18 @@ let error_spans ~(error_type : Out.core_error_kind) ~(location : Out.location) =
          spans = [dataclasses.replace(..., config_path=yaml_path)]
       *)
       let span =
+        (* This code matches the Python code.
+           Not sure what it does, frankly. *)
```
Comment (aryx) [inline src/osemgrep/cli_scan/Cli_json_output.ml:221]: This was useful for the playground I think. @emjin wrote it I think.

File: src/reporting/Semgrep_error_code.ml [hunk 5]
```diff
@@ -108,56 +154,28 @@ let known_exn_to_error ?(rule_id = None) file (e : Exception.t) : error option =
         | Tok.OriginTok { str; _ } -> spf "`%s` was unexpected" str
         | __else__ -> "unknown reason"
       in
-      Some (mk_error_tok tok msg Out.ParseError)
+      Some (mk_error_tok rule_id tok msg Out.ParseError)
   | Parsing_error.Other_error (s, tok) ->
-      Some (mk_error_tok ~rule_id tok s Out.SpecifiedParseError)
-  | R.Err err -> (
-      match err with
```
Comment (aryx) [inline src/reporting/Semgrep_error_code.ml:115]: why all of this was deleted?

**Review by aryx — APPROVED** [review]

**Comment by aryx** [comment]
Where is the PR on semgrep-interfaces?

File: semgrep.opam (hunk 1 above)
Comment (mjambon) [reply semgrep.opam]: I don't think it matters. There was a typo.

File: src/reporting/Semgrep_error_code.ml (hunk 5 above)
Comment (mjambon) [reply src/reporting/Semgrep_error_code.ml:115]: It moved to the function `opt_error_of_rule_error` above.

Comment (mjambon) [reply src/reporting/Semgrep_error_code.ml:115]: The exception was renamed `Rule.Err` -> `Rule.Error` because the other `Error` constructor occurring in another type is elegantly disambiguated with a type annotation.

In general, use this:
```
match (x : Foo.bar) with
| Error _ -> ...
```
rather than
```
match x with
| Foo.Error _ -> ...
```
because the latter can be ambiguous and misses an opportunity to tell the reader what type we're dealing with.

**Comment by mjambon** [comment]
I still have to write a changelog, write the docs, and review the whole thing myself, and in particular the changes in the output format.

File: src/core/Rule.ml (hunk 3 above)
Comment (mjambon) [reply src/core/Rule.ml:609]: yes. Eventually, I'll learn.

**Comment by mjambon** [comment]
phew. The tests pass and I'm done with fixes in the ATD interface.

**Comment by aryx** [comment]
And we forgot to test with pysemgrep ...
